docs/visibility/api_design.md

```diff
@@ -0,0 +1,62 @@
+# API Design
+
+`rafx-visibility` is a low-level crate used by `rafx-framework` for providing a `VisibilityWorld`. A `VisibilityWorld` is like a physics world -- it's a retained API that maintains a shadow state of the application's game world. The API encapsulates the game code from changes to the visibility's code algorithms or data structures. As new features are added to `rafx-visibility` like occlusion culling or hierarchical cull structures for large game scenes, those features will become available transparently to the applications relying on `rafx-visibility`.
+
+## High-level Overview
+
+### Public API
+
+The public API to `rafx-visibility` is contained in `src/*` and `src/geometry/*`.
+
+The main entry point is the `VisibilityWorldArc`. The application creates `VisibilityWorldArc` and registers new `Objects`, `Models`, `Zones`, `Volumes`, and `View Frustums` with it to mirror their game's scene. Additionally, the application can use a channel to send asynchronous commands to the `VisibilityWorldArc` in a thread-safe manner that can be processed during the next update of the `VisibilityWorld`.
+
+The `VisibilityWorld` uses handles to avoid bleeding internal design into the application.
+
+- A `ViewFrustumHandle` represents a view in the application, like the main camera, a mini-map, or a light source needed for shadow casting. It has a specific `Projection`. The supported `Projection` types are `Orthogonal` or `Perspective`.
```

Comment:
At some point we should figure out a consistent naming style for suffixes like Handle. Should the suffixes "Handle" and "Arc" have some consistent semantic meaning? (i.e. handle has to be deleted manually, arc can simply be dropped.) This is probably a larger conversation to have about rafx as a whole, not just this PR. Happy to merge this now and revisit later.

I'm also wondering if we should prefix these types with something like "Viz". (Kind of how all the low-level API types are prefixed "Rafx", I think it can help with readability.) The main concern is that sometimes names get kind of generic (like "ModelHandle" and "ZoneHandle" aren't obvious that they are visibility-related, and it's very plausible someone could have their own concept of Zone or Model in their code that uses rafx.)

Reply:
Hmm, I kind of like the idea of using `Handle` to mean "this must be deleted manually" vs `Arc` implies "you can just drop it". I'd have to go & update a bunch of the names but I don't mind.

Reply:
> I'm also wondering if we should prefix these types with something like "Viz". (Kind of how all the low-level API types are prefixed "Rafx", I think it can help with readability.) The main concern is that sometimes names get kind of generic (like "ModelHandle" and "ZoneHandle" aren't obvious that they are visibility-related, and it's very plausible someone could have their own concept of Zone or Model in their code that uses rafx.)

Isn't that what namespaces are intended to solve? In the event that scenario occurred, they could refer to the `Rafx` types using a partially or fully qualified name.

Reply:
We could do renaming in a follow-up PR if we decide on a specific plan. This is already a big PR and will get unwieldy if it sits long. (We can add issues for anything we decide to do later)

I still try to avoid name collisions because I think it makes code unnecessarily hard to read when some places use a::TypeName and other places use b::TypeName.. every time you see TypeName somewhere you have to look at what was imported and it's likely not to be consistent across files. Even if it's always fully specified, it feels a lot noisier to have to namespace everything. I also think it's valuable in this case that anyone reading code would know all the things beginning with "Viz" (or some other prefix) are closely related.

Reply:
re: offline discussion. I'll rename the higher-level visibility structs since it's awkward having them both be called `Handle`.